packages/node-core/src/indexer/dictionary.service.ts

```diff
@@ -250,45 +305,65 @@ export class DictionaryService {
       return undefined;
     }
 
-    const {query, variables} = this.dictionaryQuery(startBlock, queryEndBlock, batchSize, conditions);
+    const blockHeightSet = new Set<number>();
+    const entityEndBlock: {[entity: string]: number} = {};
+
+    let hasNextPage = true;
+    let afterCursor: string | undefined;
+
     try {
-      const resp = await timeout(
-        this.client.query({
-          query: gql(query),
-          variables,
-        }),
-        this.nodeConfig.dictionaryTimeout
-      );
-      const blockHeightSet = new Set<number>();
-      const entityEndBlock: {[entity: string]: number} = {};
-      for (const entity of Object.keys(resp.data)) {
-        if (entity !== '_metadata' && resp.data[entity].nodes.length >= 0) {
-          for (const node of resp.data[entity].nodes) {
-            blockHeightSet.add(Number(node.blockHeight));
-            entityEndBlock[entity] = Number(node.blockHeight); //last added event blockHeight
+      let _metadata: MetaData;
+
+      do {
+        const {query, variables} = this.dictionaryQuery(startBlock, queryEndBlock, batchSize, conditions, afterCursor);
+
+        const resp = await timeout(
+          this.client.query({
+            query: gql(query),
+            variables: {...variables, afterCursor},
+          }),
+          this.nodeConfig.dictionaryTimeout
+        );
+
+        for (const entity of Object.keys(resp.data)) {
+          if (entity !== '_metadata') {
+            const data = this.supportsCursor ? resp.data[entity].edges : resp.data[entity].nodes;
+            for (const item of data) {
+              const node = this.supportsCursor ? item.node : item;
+              blockHeightSet.add(Number(node.blockHeight));
+              entityEndBlock[entity] = Number(node.blockHeight);
+            }
+            if (this.supportsCursor) {
+              hasNextPage = resp.data[entity].pageInfo.hasNextPage;
+              afterCursor = resp.data[entity].pageInfo.endCursor;
+            }
           }
         }
-      }
-      const _metadata = resp.data._metadata;
+
+        _metadata = resp.data._metadata;
+
+        if (!this.dictionaryValidation(_metadata, startBlock)) {
+          return undefined;
+        }
+      } while (this.supportsCursor && hasNextPage);
```

Comment:
This isn't the right approach. The goal is to always use the following cursor unless the datasources changes. So each time `scopedDictionaryEntries` is called it should use the previous cursor if the datasources haven't changed.
This would align with my other comment that there should be no end block.